llama_index/callbacks/llama_debug.py

```diff
@@ -141,12 +148,53 @@ def get_event_time_info(
 
     def flush_event_logs(self) -> None:
         """Clear all events from memory."""
-        self._events = defaultdict(list)
+        self._event_pairs_by_type = defaultdict(list)
+        self._event_pairs_by_id = defaultdict(list)
         self._sequential_events = []
 
+    def launch(self, run_id: Optional[str] = None) -> None:
+        """Launch a run."""
+        self._trace_map = defaultdict(list)
+        self._cur_run_id = run_id
+
+    def shutdown(
+        self,
+        run_id: Optional[str] = None,
+        trace_map: Optional[Dict[str, List[str]]] = None,
+    ) -> None:
+        """Shutdown the current run."""
+        self._trace_map = trace_map or defaultdict(list)
+
+    def _print_trace_map(self, cur_event_id: str, level: int = 0) -> None:
+        """Recursively print trace map to terminal for debugging."""
+        event_pair = self._event_pairs_by_id[cur_event_id]
+        if event_pair:
+            time_stats = self._get_time_stats_from_event_pairs([event_pair])
+            indent = " " * level * 2
+            print(
+                f"{indent}|_{event_pair[0].event_type} -> ",
+                f"{time_stats.total_secs} seconds",
+                flush=True,
+            )
+
+        child_event_ids = self._trace_map[cur_event_id]
+        for child_event_id in child_event_ids:
+            self._print_trace_map(child_event_id, level=level + 1)
+
+    def print_trace_map(self) -> None:

```

Comment:
curious: why did you go with having this as a separate function, rather than just having this printed out on `shutdown`?

Reply:
Conceptually, I'd expect the handler to do things by itself after I attach it, rather than having to interact with it.

Reply:
ya just thinking that the current design is very optimized for notebook usage. But I'm trying to think about the scenario where this is used more as a "logger".

Reply:
Done!